src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs

```diff
@@ -57,59 +57,63 @@ public override int Read(Span<byte> buffer)
             if (_mode != CompressionMode.Decompress)
                 throw new InvalidOperationException(SR.BrotliStream_Compress_UnsupportedOperation);
             EnsureNotDisposed();
-            int totalWritten = 0;
 
-            OperationStatus lastResult = OperationStatus.DestinationTooSmall;
-            // We want to continue calling Decompress until we're either out of space for output or until Decompress indicates it is finished.
-            while (buffer.Length > 0 && lastResult != OperationStatus.Done)
+            int bytesWritten;
+            while (true)
             {
+                // Decompress any data we may have in our buffer.
+                OperationStatus lastResult = _decoder.Decompress(new ReadOnlySpan<byte>(_buffer, _bufferOffset, _bufferCount), buffer, out int bytesConsumed, out bytesWritten);
+                if (lastResult == OperationStatus.InvalidData)
+                {
+                    throw new InvalidOperationException(SR.BrotliStream_Decompress_InvalidData);
+                }
+
+                if (bytesConsumed > 0)
+                {
+                    _bufferOffset += bytesConsumed;
+                    _bufferCount -= bytesConsumed;
+                }
+
+                // If we successfully decompressed any bytes, or if we've reached the end of the decompression, we're done.
+                if (bytesWritten > 0 || lastResult == OperationStatus.Done)
+                {
+                    break;
+                }
+
+                // If we need more input data, read from the underlying stream.
```

Comment:
What other cases are there, aside from NeedMoreData and the ones you already handled above?

If there are other meaningful cases here, it looks like we are just going to keep looping forever.

Reply:
I'll double check tomorrow... I believe it's just these + destination too short.